internal/rpc/digest_md5_integrity.go

```diff
@@ -0,0 +1,121 @@
+package rpc
+
+import (
+	"bytes"
+	"encoding/binary"
+	"fmt"
+	"io"
+	"net"
+	"syscall"
+	"time"
+)
+
+type digestMD5IntegrityConn struct {
+	baseConnWrapper
+	buf bytes.Buffer
+
+	kic []byte
+	kis []byte
+
+	sendSeqNum int
+	readSeqNum int
+}
+
+func newDigestMD5IntegrityConn(conn net.Conn, kic, kis []byte) digestMD5Handler {
+	return &digestMD5IntegrityConn{
+		baseConnWrapper: baseConnWrapper{conn: conn},
+		kic:             kic,
+		kis:             kis,
+	}
+}
+
+func (d *digestMD5IntegrityConn) Write(b []byte) (n int, err error) {
+	if !d.writeDeadline.IsZero() && d.writeDeadline.Before(time.Now()) {
+		return 0, syscall.ETIMEDOUT
+	}
+
+	inputLen := len(b)
+	var seqBuf [4]byte
+	lenEncodeBytes(d.sendSeqNum, seqBuf[:])
+
+	outputLen := macDataLen + inputLen + macHMACLen + macMsgTypeLen + macSeqNumLen
+
+	n = 4
+	err = binary.Write(d.conn, binary.BigEndian, int32(outputLen-macDataLen))
+	if err != nil {
+		return
+	}
+
+	if n, err = d.conn.Write(b); err != nil {
+		return
+	}
+
+	hmac := getHMAC(d.kic, seqBuf[:], b)
+	if n, err = d.conn.Write(hmac); err != nil {
+		return
+	}
+
+	if n, err = d.conn.Write(macMsgType[:]); err != nil {
+		return
+	}
+
+	err = binary.Write(d.conn, binary.BigEndian, int32(d.sendSeqNum))
+	n += 4
+
+	d.sendSeqNum++
+	return
+}
+
+// Read will decode the underlying bytes and then copy them from our
+// buffer into the provided byte slice
+func (d *digestMD5IntegrityConn) Read(b []byte) (int, error) {
+	if !d.readDeadline.IsZero() && d.readDeadline.Before(time.Now()) {
+		return 0, syscall.ETIMEDOUT
+	}
+
+	n, err := d.buf.Read(b)
+	if len(b) == n || (err != nil && err != io.EOF) {
+		return n, err
+	}
+
+	var sz int32
+	binary.Read(d.conn, binary.BigEndian, &sz)
+
+	d.buf.Reset()
+	d.buf.Grow(int(sz))
+	io.CopyN(&d.buf, d.conn, int64(sz))
+
+	decoded, err := d.decode(d.buf.Bytes())
+	d.buf.Truncate(len(decoded))
+	return d.buf.Read(b[n:])
+}
+
+// decode will decode a message from the server and perform the integrity
+// protection check, removing the verification and mac data in what is returned
+// the slice returned is an alias to the buffer and must be either used or
+// copied to a new slice before calling decode again
+func (d *digestMD5IntegrityConn) decode(input []byte) ([]byte, error) {
+	inputLen := len(input)
+	if inputLen < saslIntegrityPrefixLength {
+		return nil, fmt.Errorf("Input length smaller than the integrity prefix")
+	}
+
+	var seqBuf [4]byte
+	lenEncodeBytes(d.readSeqNum, seqBuf[:])
+
+	dataLen := inputLen - macHMACLen - macMsgTypeLen - macSeqNumLen
+	hmac := getHMAC(d.kis, seqBuf[:], input[:dataLen])
+
+	seqNumStart := inputLen - macSeqNumLen
+	msgTypeStart := seqNumStart - macMsgTypeLen
+	origHashStart := msgTypeStart - macHMACLen
+
+	if !bytes.Equal(hmac, input[origHashStart:origHashStart+macHMACLen]) ||
+		!bytes.Equal(macMsgType[:], input[msgTypeStart:msgTypeStart+macMsgTypeLen]) ||
```

Comment:
Is the `[:]` necessary?

Reply:
`bytes.Equal` needs a slice, but we static initialized `macMsgType` as a `[2]byte`, so you need the `[:]` for it to be passed to `bytes.Equal`.